xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java

```diff
@@ -291,7 +291,6 @@ public void run() {
           }
           if (resourceSubscribers.get(type).isEmpty()) {
             resourceSubscribers.remove(type);
-            subscribedResourceTypeUrls.remove(type.typeUrl());
```

Comment:
We definitely want to clean up the resource. Nonces are per-resource? Then it seems when the client starts watches again the server should notice the lack of nonce. The issue might be instead that we _aren't_ cleaning up `AdsStream.respNonces`?

Note that maybe we should do the new I/O you are causing in this PR, but maybe we allow sending the ACK even when subscribedResourceTypeUrls lacks the type.

Reply:
Instead of removing line 294, have it call a cleanup method on the `subscriber.controlPlaneClient` (if it isn't null) to remove the nonce.  You'll have to create the cleanup method that you'll be calling.

Reply:
That sounds good to me. @lujiajing1126, if it turns out to be annoying to make that change, tell us and we'll see how we can help. Also, if you think that wouldn't fully address what you noticed, say so. I don't fully understand "adjustResourceSubscription issue;" it just looks like the same nonce issue to me.

Reply:
(Also, if you are uncertain about the changes, you can send them out before you update/fix any tests. A sort of early review.)

Reply:
> We definitely want to clean up the resource. Nonces are per-resource? Then it seems when the client starts watches again the server should notice the lack of nonce. The issue might be instead that we aren't cleaning up AdsStream.respNonces?

> Instead of removing line 294, have it call a cleanup method on the `subscriber.controlPlaneClient` (if it isn't null) to remove the nonce. You'll have to create the cleanup method that you'll be calling.

I agree with both of you. Instead of creating a cleanup method, I've merged cleanup logic into the existing `adjustResourceSubscription` method. PTAL

Reply:
> That sounds good to me. @lujiajing1126, if it turns out to be annoying to make that change, tell us and we'll see how we can help. Also, if you think that wouldn't fully address what you noticed, say so. I don't fully understand "adjustResourceSubscription issue;" it just looks like the same nonce issue to me.

Yes. Exactly

Reply:
> (Also, if you are uncertain about the changes, you can send them out before you update/fix any tests. A sort of early review.)

I tried to fix this issue based on the comment (without modifying the test cases)

Reply:
Please modify the failing test case to expect the nonce to be reset.

Reply:
> Please modify the failing test case to expect the nonce to be reset.

Test case has been fixed with some additional helper to access the underlying private/package-private fields